### PR TITLE
future.compat import

### PR DESCRIPTION
**Pull request by johanfylling**
Adding support for the ~`future.strict`~`future.compat` import.

If imported, we enforce for that module:

* `future.keywords.*` import is implied, and not allowed
* `if` and `contains` keywords are required when declaring rules (when applicable)
* "naked" ref-head rules without value assignment and body are disallowed
  * e.g. `a.b.c`
* the following strict-mode checks are enforced:
  * duplicate imports
  * `input` and `data` shadowing
  * deprecated built-in usage
* the following strict-mode checks are _not_ imported:
  * unused imports
  * unused local assignments

`opa fmt` will respect the ~`future.strict`~`future.compat` import and not change compliant code, but it will not correct non-compliant code.

Fixes: #6247

### CODE REVIEW

File: ast/compile_test.go [hunk 1]
```diff
@@ -3138,6 +3138,412 @@ func assertErrors(t *testing.T, actual Errors, expected Errors, assertLocation b
...
+			expectedErrors: Errors{
+				&Error{
+					Message:  "rules must not shadow input (use a different rule name)",
+					Location: &Location{Text: []byte("input := 2"), File: "policy2.rego", Row: 3, Col: 6},
+				},
+			},
+		},
+		{
+			note: "rule shadows data",
+			modules: map[string]string{
```
Comment (ashutosh-narkar) [inline ast/compile_test.go:3427]: What about a test case where a ref-head shadows data/input ?

File: ast/parser_test.go [hunk 2]
```diff
@@ -1302,6 +1333,194 @@ func TestFutureImportsExtraction(t *testing.T) {
...
+				"rego_parse_error: `if` keyword is required before rule body",
+			},
+		},
+		{
+			note: "`contains` and `if` keyword not used on partial object rule (ref-head)",
+			module: `package test
+import future.strict
+p.q[x] { x = input.x}`,
+			expectedErrors: []string{
+				"rego_parse_error: `if` keyword is required before rule body",
```
Comment (ashutosh-narkar) [inline ast/parser_test.go:1497]: For partial set rules, we get 2 errors for the `contains` and `if`. Here we get one. We would need `contains` as well, correct?

File: ast/compile_test.go (hunk 1 above)
Comment (johanfylling) [reply ast/compile_test.go:3427]: Currently, something like `data.a.b := 1` or `input.a.b := 1` doesn't generate errors. They probably should, though ..

The `input` case will cause a complete replacement of the input document: https://play.openpolicyagent.org/p/ihpFa7iM1p.

The `data` case doesn't produce error or output (!): https://play.openpolicyagent.org/p/shSDlF4mPE

I think it's out of scope to be fixed here, but we should plug these holes.

File: ast/parser_test.go (hunk 2 above)
Comment (johanfylling) [reply ast/parser_test.go:1497]: No, this is a partial object rule that has a single-value leaf, not a multi-value set, so a `contains` error isn't expected. The test case `note` is a bit confusing, though (probably a copy/paste mistake); updating.

This is one of the reasons for why we're enforcing the use of `if` and `contains` in `1.0`; today, the policy

```rego
package play

a[b] { b := 1}

x.y[z] { z := 2 }
```

will produce: 

```json
{
    "a": [
        1
    ],
    "x": {
        "y": {
            "2": true
        }
    }
}
```

For the uninitiated, it seems very random why `a` would be a set, and `x.y` would be an object.

All rules declared with `if` is an object unless `contains` is also used, though:

```rego
package play
import future.keywords

a[b] if { b := 1}

x.y[z] if { z := 2 }

```
-->
```json
{
    "a": {
        "1": true
    },
    "x": {
        "y": {
            "2": true
        }
    }
}
```
&
```rego
package play
import future.keywords

a contains b if { b := 1}

x.y contains z if { z := 2 }

```
-->
```json
{
    "a": [
        1
    ],
    "x": {
        "y": [
            2
        ]
    }
}
```
As illustrated above, enforcing the use of `if` and `contains` removes the current ambiguity.

File: ast/compile_test.go (hunk 1 above)
Comment (johanfylling) [reply ast/compile_test.go:3463]: https://github.com/open-policy-agent/opa/issues/6291

**Review by ashutosh-narkar — APPROVED** [review]
LGTM. Thanks for adding more test cases! We should update `future.strict` to `future.compat`.

**Comment by johanfylling** [comment]
@ashutosh-narkar , I've renamed the `future.strict` import to `future.compat`. Please have another look.

File: ast/parser.go [hunk 3]
```diff
@@ -2512,6 +2525,28 @@ func (p *Parser) futureImport(imp *Import, allowedFutureKeywords map[string]toke
 		kwds = append(kwds, k)
 	}
 
+	if path[1].Equal(futureCompatibleRef[1]) {
+		if len(path) > 2 {
+			p.errorf(imp.Path.Location, "invalid import, must be `%s`", futureCompatibleRef)
+			return
+		}
+		if p.s.s.HasKeyword(futureKeywords) && !p.s.s.FutureCompatible() {
+			// We have imported future keywords, but they didn't come from another `future.strict` import.
```
Comment (ashutosh-narkar) [inline ast/parser.go:2534]: Nit: `future.strict` -> `future.compat`

**Review by ashutosh-narkar — DISMISSED** [review]
LGTM

**Pull request merged**